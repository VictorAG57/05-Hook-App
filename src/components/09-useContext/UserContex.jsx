import { createContext } from 'react';

// Primero necesitamos crear un context, para ello usamos el createContext
// y lo guardamos dentro de una variable, y dentro de esá variable es con la
// que manejaremos el context en otros componentes pata compartir la información 
// entre ellos.
export const UserContext = createContext( null );