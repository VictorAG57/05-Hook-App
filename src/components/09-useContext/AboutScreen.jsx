import React, { useContext } from 'react'
import { UserContext } from './UserContex';

const AboutScreen = () => {

    const { user, setUser } = useContext( UserContext )

    return (
        <div>
            <h1>Abot Screen</h1>
        </div>
    );
};

export default AboutScreen;
