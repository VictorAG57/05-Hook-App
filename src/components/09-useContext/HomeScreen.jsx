import React, { useContext } from 'react'
import { UserContext } from './UserContex';

const HomeScreen = () => {

    // Hacemos uso del useContext, para tener acceso a la
    // información almacenada dentro de ella y tener poder
    // modificar o usar su información.
    // En esté caso usaremos la información a mutado <LoginScreen />
    // y creamos la comunicación entre componentes sin necesidad de 
    // props o que los componentes sean hijos.
    const { user } = useContext( UserContext );

    return (
        <div>
            <h1>Home Screen - Página inicial</h1>
            <hr></hr>

            <pre>
                { JSON.stringify( user, null, 3 )}
            </pre>
        </div>
    );
};

export default HomeScreen;
