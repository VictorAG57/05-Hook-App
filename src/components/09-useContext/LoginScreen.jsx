import React, { useContext } from 'react';
// Es necesario importar el context.
import { UserContext } from './UserContex';

const LoginScreen = () => {
    // Haremos una comunicación entre componentes, mediante <LoginScreen></LoginScreen>
    // cambiaremoes el estado de UserContext con un bóton, dicha información
    // se renderizara en otro componente: <HomeScreen></HomeScreen>, de está
    // manera hacemos uso de useContext con la comunicación de componentes, 
    // cambiamos el estado en una, y se renderizara en otra, toda esto sin usar
    // props, y así compartimos info sn nesecidad de que los componetes sean hijos.

    // 1. Obtener la referencia a de mi UserContext
    const { user, setUser } = useContext( UserContext );

    // 2. establecer el objeto o valor nuevo
    const user1 = {
        id: 123,
        name: 'Victor'
    };
    
    // 3. Con el siguiente bónton cambios el estado y hará que se renderize la info
    // en el componente <HomeScreen></HomeScreen>
    return (
        <div>
            <h1>Login Screen</h1>
            <hr></hr>

            <button 
                className="btn btn-primary"
                onClick={ ()=> setUser( user1 )}
            >
                Login
            </button>
        </div>
    );
};

export default LoginScreen;
