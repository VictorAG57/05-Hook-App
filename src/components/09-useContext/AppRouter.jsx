import React from 'react';

// Importaciones de React Router
import {
    BrowserRouter as Router,
    Switch,
    Route,
    Redirect,
  } from "react-router-dom"; 

// Nuestras importaciones
import AboutScreen from './AboutScreen';
import HomeScreen from './HomeScreen';
import LoginScreen from './LoginScreen';
import NavBar from './NavBar';

// Esté componente se encargara de manejar mis rutas de la librería
// React Router, de está manera cada ruta que creemos la estaremos
// viendo acá.

const AppRouter = () => {
    return (
        <Router>
            <div>
                <NavBar></NavBar>
                <Switch>
                    { /** Página inicial */}
                    <Route exact path="/" component={ HomeScreen }></Route>

                    <Route exact path="/login" component={ LoginScreen }></Route>
                    <Route exact path="/about" component={ AboutScreen }></Route>

                    <Redirect to="/"></Redirect>
                </Switch>
            </div>
        </Router>
    );
};

export default AppRouter;
