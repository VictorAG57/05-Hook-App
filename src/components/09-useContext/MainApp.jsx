import React, { useState } from 'react';
import AppRouter from './AppRouter';
// Es necesario importar nuestro context
import { UserContext } from './UserContex';

const MainApp = () => {

    // Mediante el useState oidremos cambiar la información dentro del useContext
    // y así al comunicarse entre componentes pueden manejar y modificar esá información
    // dentro del context y se actualizara en todos los componentes, haciendo
    // muy facil el cambio y comunicación de información y componentes.
    const [user, setUser] = useState({ })

    // Es necesario que una vez creado nuestro useContext, lo importemos
    // a nuestro archivo principal, en esté caso el 'MainApp.jsx', dónde 
    // envolveremos el componente por defecto de está aplicación en nuestro
    // <UserContext.Provider></UserContext.Provider>, agregamos el .Provider para
    // que funcione correctamente.
    // Esté useContext nos permitira enviar ciertas funciones, métodos, arays, etc
    // entre componentes de manera directa, sin tenerlos que enviar mediante las propiedades
    // consecutivamente, es un archivo de orden superior donde podras consultar la
    // información o datos que allás guardado.

    // value -->
    // Mediante la propiedad value podemos enviar esá información que queremos envíar
    // entre componentes. En esté caso enviaremos un objeto dentro del value,
    // nuestro estato, que contendra toda la información actual, y el setState
    // que me permitira cambiar la información del context y actualizarla.
    return (
        <UserContext.Provider value={ {
            user,
            setUser
        } }>
            <AppRouter></AppRouter>
        </UserContext.Provider>
    );
};

export default MainApp;
 