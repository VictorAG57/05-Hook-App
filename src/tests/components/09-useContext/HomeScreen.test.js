import { mount, shallow } from "enzyme";
import HomeScreen from "../../../components/09-useContext/HomeScreen";
import { UserContext } from "../../../components/09-useContext/UserContex";

describe('Pruebas en <HomeScreen />', () => {

    // Haremos como si simularamos useContext para que funcione la prueba.
    const user = {
        name: 'Victor',
        email: 'victor@gmail.com'
    };

    // mount() lo suaremos cuando necesitamos renderizar todo en general
    // en vez de shallow(), en el caso de hacer .toMatchSnapshot() cuando haces
    // pruebas con useContext usaremos mount().

    // --> Pruebas con useContext usaremos mount(), no shallow().

    // Para que nuestro wrapper funcione con useContext, es nesesario envolver 
    // dentro al componente que deseamos renderizar, dentro del useContext,
    // como si fueramos usar useContext normalmente, esto para que el snapshot
    // funcione realmente.
    const wrapper = mount( 
        <UserContext.Provider value= {{ user }}>
            <HomeScreen></HomeScreen>
        </UserContext.Provider>
    );
    test('Debe mostrarse correctamente ', () => {

        expect( wrapper ).toMatchSnapshot();
    });
});
