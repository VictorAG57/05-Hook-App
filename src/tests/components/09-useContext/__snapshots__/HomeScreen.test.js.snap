// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HomeScreen /> Debe mostrarse correctamente  1`] = `
<div>
  <h1>
    Home Screen - PÃ¡gina inicial
  </h1>
  <hr />
  <pre>
    {
   "name": "Victor",
   "email": "victor@gmail.com"
}
  </pre>
</div>
`;
